src/posix/platform/spi_interface.cpp

```diff
@@ -0,0 +1,777 @@
+    {
+        // In this case we don't have an interrupt, so we revert to SPI polling.
+        timeout = pollingTimeout;
+    }
+
+    if (mSpiTxRefusedCount)
+    {
+        struct timeval minTimeout = {0, 0};
+
+        // We are being rate-limited by the slave. This is fairly normal behavior. Based on number of times slave has
+        // refused a transmission, we apply a minimum timeout.
+        if (mSpiTxRefusedCount < kImmediateRetryCount)
+        {
+            minTimeout.tv_usec = kImmediateRetryTimeoutUs;
+        }
+        else if (mSpiTxRefusedCount < kFastRetryCount)
+        {
+            minTimeout.tv_usec = kFastRetryTimeoutUs;
+        }
+        else
+        {
+            minTimeout.tv_usec = kSlowRetryTimeoutUs;
+        }
+
+        if (timercmp(&timeout, &minTimeout, <))
+        {
+            timeout = minTimeout;
+        }
+
+        if (mSpiTxIsReady && !mDidPrintRateLimitLog && (mSpiTxRefusedCount > 1))
+        {
+            // To avoid printing out this message over and over, we only print it out once the refused count is at two
+            // or higher when we actually have something to send the slave. And then, we only print it once.
+            otLogInfoPlat("Slave is rate limiting transactions");
+
+            mDidPrintRateLimitLog = true;
+        }
+
+        if (mSpiTxRefusedCount == kSpiTxRefuseWarnCount)
+        {
+            // Ua-oh. The slave hasn't given us a chance to send it anything for over thirty frames. If this ever
+            // happens, print out a warning to the logs.
+            otLogWarnPlat("Slave seems stuck.");
+        }
+        else if (mSpiTxRefusedCount == kSpiTxRefuseExitCount)
+        {
+            // Double ua-oh. The slave hasn't given us a chance to send it anything for over a hundred frames.
+            // This almost certainly means that the slave has locked up or gotten into an unrecoverable state.
+            DieNowWithMessage("Slave seems REALLY stuck.", OT_EXIT_FAILURE);
+        }
+    }
+    else
+    {
+        mDidPrintRateLimitLog = false;
+    }
+
+    if (timercmp(&timeout, &aTimeout, <))
+    {
+        aTimeout = timeout;
+    }
+}
+
+void SpiInterface::Process(const fd_set &aReadFdSet, const fd_set &aWriteFdSet)
+{
+    OT_UNUSED_VARIABLE(aReadFdSet);
+    OT_UNUSED_VARIABLE(aWriteFdSet);
+
+    if (FD_ISSET(mIntGpioValueFd, &aReadFdSet))
+    {
+        struct gpioevent_data event;
+
+        otLogDebgPlat("Process(): Interrupt.");
+
+        // Read event data to clear interrupt.
+        VerifyOrDie(read(mIntGpioValueFd, &event, sizeof(event)) != -1, OT_EXIT_ERROR_ERRNO);
+    }
+
+    // Service the SPI port if we can receive a packet or we have a packet to be sent.
+    if (mSpiTxIsReady || CheckInterrupt())
+    {
+        // We guard this with the above check because we don't want to overwrite any previously received frames.
+        PushPullSpi();
+    }
+}
+
+otError SpiInterface::WaitForFrame(struct timeval &aTimeout)
```

Comment:
The`aTimeout` should return the time left for retrying if it fails this time.

Reply:
The RadioSpinel will calculate the  time left by itself. The function doesn't need to return the time left. The argument should be `const`, I have updated it.